Multi-Asset UTxO: Upgrade primitive types and DB schema

jonathanknowles opened the pull request:
# Issue Numbers

This PR is the first of several that implement the design described in [ADP-347](https://jira.iohk.io/browse/ADP-347) (_Wallet multi-asset UTxO_).

In particular, it implements the following subtasks:
  - [ADP-601](https://jira.iohk.io/browse/ADP-601) (_Upgrade primitive types_)
  - [ADP-602](https://jira.iohk.io/browse/ADP-602) (_Upgrade database schema_)

# User-facing changes

**None.**

# API changes

**None.**

# Overview

This section is intended to be a high-level summary.

For explanations of the design choices made by this PR, please see [ADP-347](https://jira.iohk.io/browse/ADP-347).

## Functionality that should be unaffected by this PR

<details>

### Coin Selection

The coin selection algorithms should operate in exactly the same way as they did before.

This PR modifies these algorithms in only a very small way: to look at only the `Coin` part of `TokenBundle`. Everything else should be the same.

</details>

## Types introduced by this PR

</details>

<details>
<br>

* `TokenPolicyId`

    Uniquely identifies a token policy based on the hash of its policy script.

    ```hs
    newtype TokenPolicyId = TokenPolicyId
        { unTokenPolicyId :: Hash "TokenPolicy" }
    ```

* `TokenName`

    Names a token within the scope of a policy.

    ```hs
    newtype TokenName = TokenName
        { unTokenName :: Text }
    ```

* `TokenQuantity`

    An integral quantity of tokens.

    ```hs
    newtype TokenQuantity = TokenQuantity
        { unTokenQuantity :: Integer }
    ```

* `AssetId`

    Combines a `TokenPolicyId` with a `TokenName` to uniquely identify an asset.

    ```hs
    data AssetId = AssetId
        { tokenPolicyId
            :: !TokenPolicyId
        , tokenName
            :: !TokenName
        }
    ```

* `TokenMap`

    A map of **non-ada** token quantities, scoped by `PolicyId` and `TokenName`.

    ```hs
    newtype TokenMap = TokenMap
        { unTokenMap
            :: Map TokenPolicyId (NonEmptyMap TokenName TokenQuantity)
        }
    ```

* `TokenBundle`

   Combines a non-ada `TokenMap` with an ada `Coin` quantity.

    ```hs
    data TokenBundle = TokenBundle
        { coin
            :: Coin
        , tokens
            :: TokenMap
        }
    ```

</details>

## Types modified by this PR

<details>
<br>

* `TxOut`

    ```patch
    data TxOut = TxOut
        { address
            :: !Address
    -   , coin
    -       :: !Coin
    -     -- ^ The ada amount associated with this output.
    +   , tokens
    +        :: !TokenBundle
    +      -- ^ The tokens associated with this output.
        }
    ```
</details>

## Database tables introduced by this PR

<details>
<br>

### `TxOutToken`

![TxOutToken](https://user-images.githubusercontent.com/206319/101614445-eb516080-3a47-11eb-804e-1eac7ab5a264.png)

([PDF](https://github.com/jonathanknowles/cardano-wallet/files/5664999/TxOutToken.pdf))

### `UTxOToken`

![UTxOToken](https://user-images.githubusercontent.com/206319/101614467-f2786e80-3a47-11eb-99ca-783fb088d75e.png)

([PDF](https://github.com/jonathanknowles/cardano-wallet/files/5665005/UTxOToken.pdf))

</details>

## Testing
<details>
<br>

### QuickCheck state machine and DB property tests

The existing property tests have been upgraded to generate `UTxO` and `TxOut` values that include multi-asset token bundles.

Care has been taken to ensure that we only generate token bundles of a reasonable size, to minimize our memory footprint.

### Soak testing

As part of testing this PR, the `cardano-wallet-core:test:unit` property test suite has been run hundreds of times to verify that the pass rate and total running time have not been adversely affected.

Some increase in run-time is to be expected, however, as in order to test DB serialization and deserialization, we must generate values of type `TokenBundle`, which occupy more space and take longer to generate than values of type `Coin`.

### Generators and Shrinkers

Generators and shrinkers for token-related types are located within `Gen` modules, according to the following general pattern:

- For type `T` located in module `M`, generators and shrinkers are located in module `M.Gen`.

- Generators and shinkers are defined in pairs, according to the naming pattern `{gen,shrink}TypeRange`, where `Type` is the name of the type, and `Range` refers to the range of values that are generated.

For example:

 - **`{gen,shrink}TokenBundleSmallRange`**
        Generate (or shrink) values of type `TokenBundle`, where elements of the data structure are chosen from small ranges, to promote collisions.

 - **`{gen,shrink}TokenQuantityLargePositive`**
        Generates(or shrink) large values of type `TokenQuantity` that are guaranteed to be strictly positive.

</details>

jonathanknowles committed:
Add `TxOutToken` table to wallet database.

jonathanknowles committed:
Quieten `Show` instances for certain wallet primitive types.

This change quietens the output of `Show` instances for the following types:

- `Coin`
- `Hash`

Before:

  >>> show $ Coin 1000
  Coin {getCoin = 1000}

  >>> show $ Hash "AAAA"
  Hash {getHash = "AAAA"}

After:

  >>> show $ Coin 1000
  Coin 1000

  >>> show $ Hash "AAAA"
  Hash "AAAA"

jonathanknowles committed:
Add `TokenBundle` type and related types.

jonathanknowles committed:
Add function `testLaws` for testing type class laws.

Add function `testLaws`, which can be used to test that a given instance
of a type class obeys the laws associated with that type class.

jonathanknowles committed:
Rename `getCoin` unwrapper function to `unCoin`.

This is consistent with the meaning, which is to *unwrap* a `Coin` rather than
to *get* a `Coin`.

It also prevents a clash with other functions (yet to be introduced) called
`getCoin`, which will actually get a `Coin`.

jonathanknowles committed:
Use a dedicated `coin` field within `TokenBundle` to store the ada quantity.

In this change, we:

1. Use a dedicated `coin` field within `TokenBundle` to store the ada `Coin`
   amount.

2. Extract out the map part of `TokenBundle` into a separate submodule.
   The map has an invariant that must be strictly adhered to, so it makes
   sense to limit the size of its interface (so as to limit the number of
   ways in which the invariant could be violated) and test it in isolation.

The interface of `TokenBundle` is more-or-less preserved as-is.

jonathanknowles committed:
Correct broken documentation comment for `Coin` type.

jonathanknowles committed:
Modify wallet DB to serialize token bundles in transaction outputs.

This change modifies the wallet DB SQLite implementation to serialize
token bundles held within transaction outputs to the `TxOutToken` table,
and to deserialize token bundles when deserializing transaction outputs.

jonathanknowles committed:
Correct typo in comment for `selectTxs`.

jonathanknowles committed:
Adjust selected code sections to comply with team coding standards.

jonathanknowles committed:
Add `UTxOToken` table to wallet database.

jonathanknowles committed:
Modify wallet DB to serialize token bundles in UTxO entries.

This change modifies the wallet DB SQLite implementation to serialize
token bundles held within UTxO entries to the `UTxOToken` table,
and to deserialize token bundles when deserializing UTxO entries.

jonathanknowles committed:
Generate token bundles with non-ada assets in wallet property tests.

Previously, the wallet property tests were restricted to generating token
bundles with only ada coin quantities.

This change modifies the `TxOut` arbitrary instance to use a full token
bundle generator.

jonathanknowles committed:
Prevent export of default data constructor for `TokenMap`.

This change adds a test to check that the default data constructor for
`TokenMap` is not exported.

The default data constructor for 'TokenMap' is not exported, by design,
as the internal data structure has an invariant that must be preserved
across all operations.

Exporting the default constructor would make it possible for functions
outside the 'TokenMap' module to break the invariant, opening the door
to subtle regressions.

jonathanknowles committed:
Flatten the serialized representations of `Flat TokenMap`.

This change further flattens the serialized `JSON` and `Buildable`
representations of `Flat TokenMap`, removing the `AssetId` structure
from both representations.

jonathanknowles committed:
Remove the `tokens` prefix from the `Buildable` instances of `TokenMap`.

This change makes it easier to include token maps in the `Buildable`
instances of other data structures.

jonathanknowles committed:
Add `Buildable` instances for `TokenBundle` type.

These instances simply add a `coin` field to the output produced by the
`Buildable` instances for `TokenMap`.

jonathanknowles committed:
Modify several `Buildable` instances to include tokens.

This change modifies the `Buildable` instances of `Wallet`, `UTxO`, and
`TxOut` so that they include the contents of their inner token bundles.

This is important for debugging test failures, as we sometimes need to be
able to inspect and compare the token bundles of two transaction outputs.

The output for an individual `TxOut` value is now of the following form:

```
- input: 2nd 66646130
  output:
    address: 2a0f753f...6e6a7064
    coin: 21
    tokens:
      - policy: aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
        tokens:
          - token: TokenC
            quantity: -704
          - token: TokenD
            quantity: 256
      - policy: bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb
        tokens:
          - token: TokenA
            quantity: -16
          - token: TokenB
            quantity: 72
          - token: TokenC
            quantity: 392
          - token: TokenD
            quantity: 360
```

jonathanknowles committed:
Add roundtrip DB persistence test for `TokenQuantity` type.

A token quantity can be any integral value: there are no lower or upper
bounds.

Therefore, it's especially important to check that we can serialize and
deserialize large and/or negative values.

jonathanknowles committed:
Modify foreign key constraint for `UTxOToken` to refer to `Checkpoint`.

Ideally, we'd like to define a foreign key constraint on the UTxO table,
as each row in UTxOToken refers to a unique row in UTxO.

Unfortunately, there's a bug in our version of persistent that breaks
foreign key constraints where the foreign table has a manually-defined
table name. In the case of the UTxO, the table name is "utxo", and not
"u_tx_o" (which is what the auto-generated table name would be).

To work around this bug, we instead define a foreign key constraint on
the Checkpoint table. Under the current design of rollback, this gives
equivalent behaviour, as we always delete UTxO entries by deleting their
parent checkpoints.

jonathanknowles committed:
Use explicit QC generators and shrinkers for token-related types.

This change adjusts QC generation code for token-related types so that:

  * Each `Arbitrary` instance is defined in terms of a named generator and
    named shrinker, as a matching pair: `(genX, shrinkX)`.

  * Shared generators and shrinkers corresponding to a given type located
    in module `M` are located in module `M.Gen`.

  * We no longer import `Arbitrary` instances from one test module to another.

Advantages:

  * It's easier to re-use generators and shrinkers in a controlled fashion, as
    we can import them without also importing a set of `Arbitrary` instances
    "for free".

  * It's easier to understand how each `Arbitrary` instance is defined, as we
    can just navigate directly to the definition of each generator and shrinker
    rather than trying to follow a chain of imported `Arbitrary` instances,
    which is time consuming and error-prone.

  * It helps us to avoid the "spooky action at a distance" problem, where
    changing the definition of an `Arbitrary` instance in one test module ends
    up affecting the coverage of a property in another test module.

Disadvantage:

  * To make it possible for both `cardano-wallet-core` and `cardano-wallet` to
    consume these generators and shrinkers, we define each `Gen` module within
    the `cardando-wallet-core` package.

    This requires `cardano-wallet-core` to include test generation code and
    have a dependency on `QuickCheck`, which is not ideal. In future, we should
    move these generators to a separate library.

jonathanknowles committed:
Rescue difficult-to-read code section in `FeeSpec`.

Later commits will modify this part of the code, so this commit tidies it
up pre-emptively.

This change:

- Replaces all instances of `Coin {getCoin = n}` with `Coin n`.
- Replaces all instances of `Address {getAddress = a}` with `Address a`.
- For multi-line records, places each field at the same indent level.
- Defines some constants and uses them to simply nested sub-expressions.

The resulting code should be equivalent, but easier to read and modify.

jonathanknowles committed:
Replace `Coin` with `TokenBundle` in `TxOut`.

This change adjusts the `TxOut` type to include a `TokenBundle` rather
than just a `Coin`. By extension, `UTxO` values are now capable of
holding multiple different token quantities.

This change also makes a small modification to preserve the current
behaviour of coin selection, so that it only considers the ada coin
value of a `TxOut` rather than considering the whole token bundle. This
state of affairs is temporary: future changes will upgrade coin
selection to consider all the tokens in a bundle. This might require
fundamental changes to the way that our coin selection algorithms work.

Some care was required to make sure that our property tests still
continue to work in the face of this change. In particular, for our coin
selection tests, we currently only generate `TxOut` values with
singleton ada token bundles. This is also temporary: future changes will
upgrade our coin selection generators to generate multi-asset token
bundles.

jonathanknowles committed:
Move `TokenMap` type out of `TokenBundle` module hierarchy.

The `TokenMap` type, apart from being a sub-component of `TokenBundle`,
is useful in its own right.

When the wallet eventually adds support for the `mint` field of a
transaction (which will encode tokens the transaction is putting into or
taking out of circulation), the wallet will need a way to represent
token bundles that cannot hold ada quantities.

The `TokenMap` type satisfies this use case, as it can only hold non-ada
token quantities.

jonathanknowles committed:
Clarify module documentation for `TokenBundle` and `TokenMap`.

jonathanknowles committed:
Regenerate JSON golden test files.

As a result of restructuring the QC generators and shrinkers for `Coin`
and token-related types, the same starting seed generates different JSON
examples for several types.

This commit adds updated JSON golden test files for these types.

jonathanknowles commented on lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs:
```diff
@@ -263,6 +267,33 @@ instance PathPiece TxId where
+
+instance PersistField TokenQuantity where
+    toPersistValue = toPersistValue . toText
+    fromPersistValue = fromPersistValueFromText
```
Perhaps add a comment here to mention why we are serializing to text.

(Neither SQLite nor `persistent-sqlite` support a big integer data type.)

jonathanknowles replied:
Fixed in b76899d31d28b0d6342eea43e28f9e399b00c55b.

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -1655,7 +1656,10 @@ instance Arbitrary RewardAccount where
+    arbitrary = genStrictlyPositiveCoin
+
+genStrictlyPositiveCoin :: Gen Coin
+genStrictlyPositiveCoin = Coin <$> choose (1, 1_000_000_000_000_000)
```
Perhaps we can replace this with calls to `Cardano.Wallet.Primitive.Types.Coin.Gen.genCoinLargePositive`.

jonathanknowles replied:
Fixed in bce20553d283b5a13c5eeccb76e19d6f88aa7e16.

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:
```diff
@@ -411,7 +413,10 @@ instance Arbitrary TxMetadata where
+    arbitrary = genStrictlyPositiveCoin
+
+genStrictlyPositiveCoin :: Gen Coin
+genStrictlyPositiveCoin = Coin <$> choose (1, 100_000)
```
Again, we should consider replacing this with calls to `genCoinLargePositive` (or similar).

jonathanknowles replied:
Fixed in bce20553d283b5a13c5eeccb76e19d6f88aa7e16.

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelectionSpec.hs:
```diff
@@ -375,7 +377,10 @@ instance Arbitrary Address where
+    arbitrary = genStrictlyPositiveCoin
+
+genStrictlyPositiveCoin :: Gen Coin
+genStrictlyPositiveCoin = Coin <$> choose (1, 100_000)
```
Again, replace this a call to a generator from `Coin.Gen`.

jonathanknowles replied:
Fixed in bce20553d283b5a13c5eeccb76e19d6f88aa7e16.

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -452,10 +456,11 @@ instance Arbitrary (WithPending WalletState) where
-                        , resolvedInputs = [(inp, coin out)]
-                        , outputs = [out { coin = simulateFee (coin out) }]
+                        , resolvedInputs = [(inp, txOutCoin out)]
+                        , outputs = [out {tokens}]
```
This needs checking. It looks like the call to `simulateFee` was missed out here.

jonathanknowles replied:
Ah, turns out it wasn't missed out after all:

```hs
                let tokens = TB.fromCoin $ simulateFee $ txOutCoin out
                let pending = withWithdrawal $ Tx 
                        { txId = arbitraryHash
                        , resolvedInputs = [(inp, txOutCoin out)] 
                        , outputs = [out {tokens}]
                        , withdrawals = mempty
                        , metadata = Nothing  
                        }
```

jonathanknowles commented on lib/core/test/unit/Data/Map/NonEmpty/StrictSpec.hs:
```diff
@@ -0,0 +1,199 @@
+            it "prop_insert" $
+                property prop_insert
+            it "prop_delete" $
+                property prop_delete
```
Looks like we're missing a `prop_delete_invariant` here.

jonathanknowles replied:
It turns out we don't need an invariant test here, as the type of `delete` is `k -> NonEmptyMap k a -> Map k a`. The result is an ordinary map.

jonathanknowles committed:
Remove `subtract` operation from `TokenMap`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2384/files#r539923648

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types/TokenMap.hs:
```diff
@@ -0,0 +1,539 @@
+subtract a b = F.foldl' acc a $ toFlatList b
+  where
+    acc c (asset, quantity) =
+        adjustQuantity c asset (`TQ.subtract` quantity)
```
After reviewing with @rvl, we decided to remove the `subtract` operation for the moment, as it's not clear that it gives us anything of value.

If we need to add this back in later, we can.

jonathanknowles replied:
Fixed in 5bbd9b34a13a5b5d0b517fd9b2d19e3a1b9d73df.

rvl commented on lib/core/cardano-wallet-core.cabal:
```diff
@@ -181,6 +199,19 @@ library
       Network.Wai.Middleware.ServerError
       Network.Wai.Middleware.Logging
       Ouroboros.Network.Client.Wallet
+        -- TODO:
```
Good.

rvl commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -763,8 +767,9 @@ newDBLayer trace defaultFieldValues mDatabaseFile ti = do
                 insertCheckpoint wid cp <* writeCache wid (Just cp)
-                let (metas, txins, txouts, ws) = mkTxHistory wid txs
-                putTxs metas txins txouts ws
+                let (metas, txins, txouts, txoutTokens, ws) =
```
What about renaming txoutTokens → txoutToks rather than breaking the line?

jonathanknowles replied:
> What about renaming txoutTokens → txoutToks rather than breaking the line?

As we already use `token` and `tokens` in other places, my preference would be to keep this name consistent with the usage in other places.

We _could_ rename `txoutTokens` to just `tokens`, but I think it's also nice for the name to indicate that these tokens are related to `txouts`.

rvl replied:
OK.

rvl commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -1281,11 +1310,13 @@ txHistoryFromEntity ti tip metas ins outs ws =
                 getTxId txid
             , W.txInfoInputs =
-                map mkTxIn $ filter ((== txid) . txInputTxId . fst) ins
+                mkTxIn <$> filter ((== txid) . txInputTxId . fst) ins
```
Was this a hlint suggestion? It doesn't seem particularly necessary - especially since the old form is still used a few lines down.

jonathanknowles replied:
> Was this a hlint suggestion?

It _might_ have been 🤷🏻. (Though I can't reproduce it -- perhaps I implemented an `hlint` suggestion at some point and then made a later change which rendered the previous suggestion invalid...)

I must admit my natural instinct is to always use `fmap` (or `<$>`). Since we also use the `a <$> b` pattern quite a bit in the code, it seemed natural to write that.

Having said that, do you have a strong preference for `map` over `fmap` in this case? Is there a reason for preferring `map`? (Perhaps to make sure that it can't be used with some other `Functor`?)

rvl replied:
Either is fine by me, but if it doesn't need to be changed, there's no need to touch it.

jonathanknowles replied:
@rvl 

Fixed in d1e2ab04151ee25f907b64b9e174245ac0d895f7.

rvl commented on lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:
```diff
@@ -52,6 +52,8 @@ import qualified Cardano.Wallet.Primitive.AddressDiscovery.Sequential as W
 import qualified Cardano.Wallet.Primitive.Types.Address as W
 import qualified Cardano.Wallet.Primitive.Types.Coin as W
+import qualified Cardano.Wallet.Primitive.Types.TokenPolicy as TP
+import qualified Cardano.Wallet.Primitive.Types.TokenQuantity as TQ
```
Just the `W` qualifier would be fine, like the rest.

jonathanknowles replied:
@rvl Fixed in af1d54eab6d521ba0d7493688fb03dc152ece007.

rvl replied:
Thanks

rvl commented on lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs:
```diff
@@ -263,6 +267,35 @@ instance PathPiece TxId where
+    sqlType _ = sqlType (Proxy @Text)
+
+instance PersistField TokenQuantity where
+    -- The `persistent-sqlite` package has no big number type, so we use a
```
It would be more correct to say that sqlite has no big integer type.

jonathanknowles replied:
@rvl wrote:
> It would be more correct to say that sqlite has no big integer type.

Good point!

I've updated the comment to reflect this.

Fixed in 9b79407ac74f5064242ffcef8dd5a40fd5e3e540.

rvl replied:
Thanks

rvl commented on lib/core/src/Cardano/Wallet/Primitive/Types/TokenBundle.hs:
```diff
@@ -0,0 +1,285 @@
+--
+-- >>> import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TB
+--
+module Cardano.Wallet.Primitive.Types.TokenBundle
```
Please would you consider following my suggestion from yesterday and rolling TokenQuantity and TokenMap into this module - TokenBundle. Because we don't need so many little modules. Pretty much the only place where TokenQuantity and TokenMap are used is TokenBundle. Think about the person who needs to read this code. If you believe splitting it up will make compile times faster, then prove it.

jonathanknowles replied:
@rvl wrote:
> rolling TokenQuantity and TokenMap into this module - TokenBundle
> If you believe splitting it up will make compile times faster, then prove it.

Compilation time is not the major motivation for having separate modules here.

`TokenMap` and `TokenBundle` are separate modules for the following reasons:

1. They fulfill different purposes:

    a. `TokenMap` defines a map of **_non-ada_** quantities, for situations where ada quantities are forbidden (such as the `mint` field of a transaction).

    b. `TokenBundle` adds an ada quantity and defines a set of extra operations for converting between bundles and `Coin` values.

2. `TokenMap` has an **invariant**. By encapsulating this in its own module and not exporting the public constructor:

    * we can limit the number of operations that potentially violate the invariant, requiring fewer tests.
    * we can avoid the complication of considering `Coin` values when writing tests for `TokenMap`, as it can't include them.<br><br>

3. `TokenMap` and `TokenBundle` have some common operations (for example, `getQuantity` and `adjustQuantity`) but they operate on different types.

    Moving them into the same module means we'd need a way to distinguish between these functions. We could use names that include the type they operate on, such as  `adjustTokenMapQuantity` and `adjustTokenBundleQuantity`, but this seems awkward. By putting them in different modules, they can have separate namespaces, removing the clash.

@rvl wrote:
> so many little modules

I'm not sure I agree that `TokenBundle` or `TokenMap` are small modules:

| Module | Size |
| -- | -- |
| `TokenMap` | 521 lines |
| `TokenBundle` | 285 lines |

I agree that `TokenQuantity` is relatively small, but it's not smaller than `Coin`, which is also separate. Because `TokenQuantity` is used by both `TokenMap` and `TokenBundle`, there doesn't seem to be a good reason to choose one or the other, unless we allow re-exports, which many people dislike due to code navigation issues.

> Think about the person who needs to read this code.

If a person is looking for the definition of type `TokenQuantity` (for example), isn't it easier to first look for a module called `TokenQuantity`, rather than trying to guess which of several other modules it could be in?

With editors that support go-to-definition, this becomes a moot point anyway.

rvl replied:
Even with go-to-definition support in your editor (even github may support it soon too), having less files is easier to work with.

Yes TokenMap and TokenBundle are different things, but is TokenMap ever going to be used independently of TokenBundle?

It's just lots of lines of fluff.

jonathanknowles replied:
@rvl wrote:
> Yes TokenMap and TokenBundle are different things, but is TokenMap ever going to be used independently of TokenBundle?

**Yes.** I actually answered that in my [earlier comment](https://github.com/input-output-hk/cardano-wallet/pull/2384#discussion_r542176792):

> `TokenMap` defines a map of non-ada quantities, for situations where ada quantities are forbidden (**such as the `mint` field of a transaction**).

Users have a choice between:
* `TokenBundle`
   allows ada quantities (used by `TxOut.tokens`).
* `TokenMap`
   does not allow ada quantities (will be used by `Tx.mint`, once we support minting).

@rvl wrote:
> Even with go-to-definition support in your editor (even github may support it soon too), having less files is easier to work with.

I'm sorry, I'm not really persuaded by this argument. If that were true, why not just put everything in one large file?

Of course, there has to be a balance somewhere.

The fact that `TokenMap` has an invariant is a good reason to encapsulate this type in its own module. Not exporting the default constructor means that callers are forced to use smart constructors, which we test in `TokenMapSpec`. 

A side benefit is that tests for `TokenBundle` don't have to care about preserving the invariant of `TokenMap`, because the responsibility is encapsulated in that module (and its tests).

I could write more but I'd just be repeating my [earlier comment](https://github.com/input-output-hk/cardano-wallet/pull/2384#discussion_r542176792).

> It's just lots of lines of fluff.

I'm not sure how this comment really helps us move forward.

Can we concentrate on specific points? I'm happy to discuss any issues you find, and make changes, so long as there are valid reasons for doing so.

rvl replied:
If you want to dig in and throw around "moving forward", then there's not much I can do, other than say that I'm trying to help you keep things simple.

Perhaps the data type for mint/burn ought to be different to the one for UTxO value? One of these is non-negative. You could consider removing the TokenQuantity module (fluff) and parameterise TokenMap by Integer or Natural. Up to you.

jonathanknowles replied:
> If you want to dig in and throw around "moving forward", then there's not much I can do, other than say that I'm trying to help you keep things simple.

First of all, I appreciate that you're trying to help. Reviewing a large PR is not easy.

But surely you must realize that comments like "It's just lots of lines of fluff" are not helpful. This comes across as a bit condescending, especially when someone has put thought and effort into a design. If you have doubts about a particular design choice, why not just reach out to me and ask? I'm happy to answer questions about anything. I'm also happy to change things (within reason) if there's a clear advantage to doing things another way.

I'd really like to overcome this hurdle and work with you on this.

When you asked me to justify the separation of `TokenMap` and `TokenBundle` (which is mentioned in the design), I gave several reasons, in good faith:

1. Encapsulating the invariant of `TokenMap`.
2. Tests for `TokenMap` can focus on properties of the map (including the invariant), without the additional complication of ada `Coin` values.
3. Separation of concerns and responsibilities. `TokenMap` is responsible for the map structure, and `TokenBundle` is responsible for interoperability with `Coin` values.
4. Allowing the functions of each type to have their own separate namespaces to avoid name clashes (if you want an analogy, think of `Map` and `NonEmptyMap`, which share similar sets of operations, but operate on different types.).

But as far as I can tell, you've not answered these points. It's exhausting to participate in a review where the reviewer doesn't respond to points raised.

This whole thread started with:
> Please would you consider following my suggestion from yesterday and rolling TokenQuantity and TokenMap into this module - TokenBundle. Because we don't need so many little modules.

I get that you have a preference for larger modules, which I respect. But `TokenMap` and `TokenBundle` are not small modules.

After some to and fro, you mentioned that:
> having less files is easier to work with.

This seems very subjective. Where do you draw the line?

I think the following are far more important than module size:

- encapsulation of invariants.
- prevention of regressions.
- separation of concerns.
- clear name-spacing.

Should we sacrifice any of these just for the sake of dropping our module count by one?

> Perhaps the data type for mint/burn ought to be different to the one for UTxO value? One of these is non-negative. You could consider removing the TokenQuantity module (fluff) and parameterise TokenMap by Integer or Natural. Up to you.

This is actually a good point. Though I don't understand the need to use antagonistic language like "fluff". Why is that necessary? As for choice to put `TokenQuantity` in a `newtype`: this was in the design document.

**As for parameterization**: I agree we should at least consider this. I propose that the right time to discuss this will come when we design support for minting. For now, we're probably not in danger of creating negative quantities, as nothing in the wallet creates real token bundles (yet). Perhaps we can settle on `Integer` or `Natural` for now, and then parameterize later, if we need to?

rvl replied:
It's not worth my time to rebut every dubious point, if I can see that you feel very strongly about this minor issue. Your PR - do as you wish.
For all this talk of invariants, the only invariant for TokenMap is that values are non-zero - but what happens if that is ever violated? Nothing probably! The wallet will very likely continue to function as normal. That invariant may even make implementing coin selection more difficult. And yet there is no invariant preventing negative token values!? This is a priority inversion due to overfluffing the fluff with too much fluff. Fluff is a technical term meaning low function to LOC ratio.

jonathanknowles replied:
> the only invariant for TokenMap is that values are non-zero - but what happens if that is ever violated?

The `Eq` instance will become incorrect, and it might even be the case that the minimum ada calculation (provided by the ledger) will give incorrect results (though I would need to double-check this).

Of course, we could rewrite our `Eq` instance to handle zero values, but that seems pointless given we have an auto-derived one that already works.

> The wallet will very likely continue to function as normal.

The wallet will also function as normal if the `TokenBundle` and `TokenMap` code is organized in 2 modules rather than 1, right? Merging modules (which was the reason for this thread) creates extra work here.

> That invariant may even make implementing coin selection more difficult. 

Can you say why the invariant would make coin selection more difficult?

The `getQuantity` function returns `0` in the case where there is no quantity for a given token, so callers shouldn't need to do anything special in the case where a bundle doesn't have a particular token.

> And yet there is no invariant preventing negative token values!?

AFAIK, this doesn't affect the integrity of the map, or the `Eq` instance.

Both [`cardano-ledger-specs`](https://github.com/input-output-hk/cardano-ledger-specs/blob/6e30ecf73c9296bc320d71f3a6b7bfad3ce117e9/shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs#L104) and [`cardano-node`](https://github.com/input-output-hk/cardano-node/blob/e20559ba6288d09bdc22a44cba7806ff85111274/cardano-api/src/Cardano/Api/Value.hs#L105) also use `Integer` here, presumably so that these types can be reused in the context of minting and burning.

But as you pointed out, we can parameterize this in a later PR, so that:
- we can use `Natural` in the case of a `TokenBundle` that's included in a transaction output.
- we can use `Integer` in the case of a non-ada `TokenMap` included in the `mint` field of a transaction.

KtorZ commented on lib/core/cardano-wallet-core.cabal:
```diff
@@ -173,6 +189,8 @@ library
       Data.Binary.Get.Safe
       Data.Function.Utils
+      Data.Map.NonEmpty.Strict
+      Data.Map.NonEmpty.Strict.Internal
```
Have you considered existing hackage / stackage alternatives? There's https://hackage.haskell.org/package/nonempty-containers although it is lazy-only. 

Beside, may I consider putting this in a separate **package**. This is decoupled enough from the wallet to live out of it.

jonathanknowles replied:
@KtorZ wrote:
> Have you considered existing hackage / stackage alternatives? There's https://hackage.haskell.org/package/nonempty-containers although it is lazy-only.

I did indeed look through Hackage for alternatives. There are several non-empty maps, but sadly they all seem to be implemented in terms of lazy maps.

> Beside, may I consider putting this in a separate package. This is decoupled enough from the wallet to live out of it.

We could do that if you prefer. How about `strict-non-empty-containers`? (Unfortunately, `non-empty-containers` is already used.)

Having said that, we don't really have a need for it outside of `cardano-wallet-core`, so we probably wouldn't gain too much from moving it (yet).

KtorZ replied:
I am not saying we should **publish** a new package, but simply organize our internal code differently. Since this is clearly not wallet-specific, there's no reason to keep it in `cardano-wallet-core`; it'll prevent the temptation from adding wallet-specific things to it.

jonathanknowles replied:
> I am not saying we should publish a new package, but simply organize our internal code differently.

Oh definitely! I didn't think that was your suggestion. Sorry if my comment gave that impression. 😄 

I was just trying to come up with an internal package name. I can use something like `strict-non-empty-containers`, unless you feel strongly about an alternative name?

jonathanknowles replied:
@KtorZ I've moved this type (and its tests) to a new internal package `strict-non-empty-containers`.

KtorZ commented on lib/core-integration/src/Test/Integration/Framework/DSL.hs:
```diff
@@ -494,7 +495,7 @@ expectWalletUTxO coins = \case
         let constructUtxoEntry idx c =
                 ( TxIn (Hash "ARBITRARY") idx
-                , TxOut addr (Coin c)
+                , TxOut addr (TB.fromCoin $ Coin c)
```
Minor suggestion because `TB` feels rather cryptic first reading it; what about `Bundle` ? `Bundle.fromCoin` reads without ambiguity.

jonathanknowles replied:
> Minor suggestion because TB feels rather cryptic first reading it; what about Bundle? Bundle.fromCoin reads without ambiguity.

@KtorZ I also like your suggestion of `Bundle`, but it raises the question of how to import other modules like `TokenMap` in a consistent way.

The PR currently import token-related modules in the following standard ways:
```
import qualified Primitive.Types.TokenBundle   as TB
import qualified Primitive.Types.TokenMap      as TM
import qualified Primitive.Types.TokenQuantity as TQ
```
We could make this a bit more verbose:
```
import qualified Primitive.Types.TokenBundle   as TBundle
import qualified Primitive.Types.TokenMap      as TMap
import qualified Primitive.Types.TokenQuantity as TQuantity
```
Or even just use the whole name:
```
import qualified Primitive.Types.TokenBundle   as TokenBundle
import qualified Primitive.Types.TokenMap      as TokenMap
import qualified Primitive.Types.TokenQuantity as TokenQuantity
```

Would any of these options work for you?

KtorZ replied:
I'd actually merge those modules together as `import qualified Primitive.Types.TokenBundle` and import it qualified as `Bundle` :no_mouth: 

The modules are very closely related and actually sharing data-structures from one another. A change in `TokenQuantity` will likely induce a change in `TokenMap` and then `TokenBundle`, so it makes sense to keep them close. Although it may not play well with the modules API which maybe overlap a bit. 

From the choices you've pointed out, the last one (whole name) is what reads clearer in my opinion. And in the end roughly the same to type. `To<tab>` will do in most cases :)

jonathanknowles replied:
> `import qualified Primitive.Types.TokenBundle   as TokenBundle`
> `import qualified Primitive.Types.TokenMap      as TokenMap`
> `import qualified Primitive.Types.TokenQuantity as TokenQuantity`

@KtorZ wrote:
> From the choices you've pointed out, the last one (whole name) is what reads clearer in my opinion. And in the end roughly the same to type. `To<tab>` will do in most cases :)

Okay!

I've fixed this in ca948b4ed9046e9e6030c29a1f7055c596aab0e7.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -1458,40 +1510,67 @@ deleteDelegationCertificates wid filters = do
+                    [ TxOutTokenTxId ==. txOutputTxId out
+                    , TxOutTokenTxIndex ==. txOutputIndex out
+                    ]
+                    []
```
The part fetching the output + token is identical (modulo the list of tx ids) here and above. I'd suggest factoring out the common part in a dedicated function to make avoid duplication and clearly show the commonality. 

On a different note, this number of requests that this generates seems rather _large_. I haven't yet checked the database benchmarks but we should make sure to measure this correctly and generate bench database with token bundles.

jonathanknowles replied:
> The part fetching the output + token is identical (modulo the list of tx ids) here and above. I'd suggest factoring out the common part in a dedicated function to make avoid duplication and clearly show the commonality.

Well spotted! I didn't notice that.

Fixed in a726e950c6530adc8dbbb5c98a039447035d7c5a.

jonathanknowles replied:
> On a different note, this number of requests that this generates seems rather large. I haven't yet checked the database benchmarks but we should make sure to measure this correctly and generate bench database with token bundles.

Great point.

The DB benchmark currently generates ada-only bundles.

I'll upgrade this to generate bundles that include non-ada tokens.

Anecdotally, when running the DB property tests, I do see a small lengthening in the total run-time (around 5% to 10%). However, this could be partly due to the fact that generating token bundles is inherently more time-consuming than just generating coins.

jonathanknowles replied:
@jonathanknowles wrote:
> I'll upgrade this to generate bundles that include non-ada tokens.

Just for the record, we decided to handle the upgrade of the DB benchmarks in a separate ticket, which I have created here: https://jira.iohk.io/browse/ADP-627

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/Types/TokenMap.hs:
```diff
@@ -0,0 +1,529 @@
+toNestedList
+    :: TokenMap -> [(TokenPolicyId, NonEmpty (TokenName, TokenQuantity))]
+toNestedList =
+    fmap (fmap NEMap.toList) . Map.toList . unTokenMap
```
Note that, when it comes to database serialization functions, I think it is better to define functions such as fold and traverse to avoid effectively traversing maps twice (once for turning it into list, and then traversing the list to serialize to the database). Although, I imagine that `TokenMap` would remain pretty small (limited by the tx max size anyway) and thus it should be _okay_.

jonathanknowles replied:
> I think it is better to define functions such as fold and traverse to avoid effectively traversing maps twice (once for turning it into list, and then traversing the list to serialize to the database).

@KtorZ I like this idea very much. Though implementing it might require quite a few changes to plumbing in the DB layer.

Would you mind at all if I were to attempt this in a follow-up PR?

I'm also interested to test the effect on the DB benchmarks, if I can.

KtorZ replied:
> Would you mind at all if I were to attempt this in a follow-up PR?

Nope, and it was more of a gratuitous remark / reminder. I don't think it'd matter much for token maps because maps would be rather small anyway (bounded by the transaction max size).

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/Types/TokenPolicy.hs:
```diff
@@ -0,0 +1,96 @@
+    fromText = fmap TokenPolicyId . fromText
+
+mkTokenPolicyId :: ByteString -> TokenPolicyId
+mkTokenPolicyId = TokenPolicyId . Hash
```
This is arguably the unsafe way to do this. I'd expect the safe constructor to be `TokenPolicy -> TokenPolicyId`

jonathanknowles replied:
> This is arguably the unsafe way to do this. I'd expect the safe constructor to be TokenPolicy -> TokenPolicyId

@KtorZ Thanks for spotting this. I had written this function very early on, and intended to make it safer later on.

I've now removed the function entirely.

Since we don't yet have a `TokenPolicy` type, I'll leave the creation of a `TokenPolicy -> TokenPolicyId` function to a future PR.

Fixed in f7d596eb3ab4920b4b9edb2c6aff4797be8b9182.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Primitive/Types/TokenBundleSpec.hs:
```diff
@@ -0,0 +1,43 @@
+    describe "Class instances obey laws" $ do
+        testLawsMany @TokenBundle
+            [ eqLaws
+            , semigroupLaws
```
Interesting package / methods :+1:

jonathanknowles replied:
I agree, [`quickcheck-classes`](https://hackage.haskell.org/package/quickcheck-classes/docs/Test-QuickCheck-Classes.html) is certainly a useful package!

I was motivated to look for this functionality after having to define the following instance:

```hs
-- Since the 'TokenBundle' type deliberately does not provide an 'Ord' instance
-- (as that would lead to arithmetically invalid orderings), this means we can't
-- automatically derive an 'Ord' instance for the 'TxOut' type.
--
-- Instead, we define an 'Ord' instance that makes comparisons based on the list
-- representation of a 'TokenBundle'.
--
instance Ord TxOut where
    compare = comparing projection
      where
        projection :: TxOut ->
            ( Address
            , Coin
            , [(TokenPolicyId, NonEmpty (TokenName, TokenQuantity))]
            )
        projection out =
            ( out & view #address
            , out & view (#tokens . #coin)
            , out & view (#tokens . #tokens) & TM.toNestedList
            )
```

KtorZ reviewed (commented):
I started (50/75) doing mainly a "surface" review to get acquainted with the changes. Nice work, I'll carry on with the review later today and early next week.

rvl commented on lib/core/src/Cardano/Wallet/Primitive/Types/TokenMap.hs:
```diff
@@ -0,0 +1,529 @@
+-- serialization style, where token quantities are paired with their asset
+-- identifiers.
+--
+newtype Flat a = Flat { getFlat :: a }
```
Are both of Flat and Nested _really_ needed?
I would expect the JSON serialisation to be defined only in the Api.Types module, and according to the swagger spec.

jonathanknowles replied:
> Are both of Flat and Nested really needed?

Good point.

We could easily get rid of one if we can show that only one is useful.

`Nested` is useful for debugging, as it groups tokens together by policy, saving space. It might also be useful for clients such as Daedalus, to provide a view that groups tokens by policy (since tokens can share the same name). But we should check this before deciding what to eliminate.

`Flat` might be preferable for clients that don't care about grouping, because it's simpler to parse.

> I would expect the JSON serialisation to be defined only in the Api.Types module

I had anticipated that these types (and their JSON instances) would be re-used by several types within the `Api.Types` module.

Though for the `TokenMap` `FromJSON` instance, we perform validation to guard against zero values, which is something that `TokenMap` itself is responsible for.

We could move these to the `Api.Types` module, but then the responsibility for guarding against zero values would be located in two modules, rather than one. It seems better to put all the validation for a type together with the type, as this way they can also be tested together.

It's actually not the first time we've done this. The following types all define JSON instances in their own modules, in some cases for similar reasons:

- `Quantity`
- `Percentage`
- `StakePoolTicker`
- `PoolOwner`
- `PoolMetadataSource`
- `SMASHPoolId`

rvl replied:
Can we try to keep the code limited to what is essential for purpose, rather than adding functions which "might be needed"?
For a start, that would make less code to review, and therefore less possible objections from me.
But also it helps reduce confusion when it comes time to refactor.

jonathanknowles replied:
> Can we try to keep the code limited to what is essential for purpose, rather than adding functions which "might be needed"?

As mentioned above, I'm  happy to remove one of these formats.

My original motivation for the flat/nested distinction was testing: visual feedback of grouping + flattening was useful while implementing these types.

I've asked for feedback from the Daedalus team if they have any strong preferences about which JSON format they'd prefer. Once we know the outcome, we can delete what we don't need.

rvl commented on lib/core/src/Cardano/Wallet/Primitive/Types/TokenMap.hs:
```diff
@@ -0,0 +1,529 @@
+-- If the given quantity is zero, the resultant map will not have an entry for
+-- the given asset.
+--
+setQuantity :: TokenMap -> AssetId -> TokenQuantity -> TokenMap
```
I just grepped for this symbol, and couldn't find where it was used by the code. Is it used by the code?

jonathanknowles replied:
> I just grepped for this symbol, and couldn't find where it was used by the code. Is it used by the code?

**EDIT:** Yes. This function is used by `adjustQuantity`, which is used by `fromFlatList`, which is used by the DB deserialization functions.

~~You're right that this function isn't currently used.~~
> ~~@rvl if you prefer, I can remove it and re-add it in a later PR that addresses coin selection?~~

rvl replied:
Yes please - I am looking forward to reviewing this function in the _later_ PR.

jonathanknowles replied:
@rvl My apologies: I had forgotten that `setQuantity` is used by `adjustQuantity`.

That function is used by `add` and `fromFlatList`, which is used by the DB deserialization functions.

So it's not really possible to remove without removing functionality that we need.

rvl replied:
OK.

rvl commented on lib/core/src/Cardano/Wallet/Primitive/Types/TokenPolicy/Gen.hs:
```diff
@@ -0,0 +1,58 @@
+mkTokenPolicyId
+    = TP.mkTokenPolicyId
+    . unsafeFromHex
+    . B8.replicate tokenPolicyIdHexStringLength
```
It might be nicer for QC counterexamples if these hashes are just single characters in a..z, or something like that?

jonathanknowles replied:
@rvl I did consider using just a single character, but the `FromText` instance for `TokenPolicyId` is based on the `Hash` type, which expects to find a byte string of the correct length. If we change the existing generator to use just one character, then the textual serialization tests will fail.

The generator and shrinker currently produce values like this:
* `aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa`
* `bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb`

But I can see it would be nicer to have:
* `PolicyA`
* `PolicyB`
  ... and so on.

Of course, we **_could_** split this generator up into two different generators:

1. Generate valid-length hashes for serialization tests.
2. Generate short hashes for all other tests.

Or, we could use `ShowFmt` in the tests, and just display the first `n` characters in the event of a test failure?

rvl replied:
OK, that's too bad then. I have found that full-length hashes make counterexamples more difficult to understand.

jonathanknowles replied:
> OK, that's too bad then. I have found that full-length hashes make counterexamples more difficult to understand.

I agree with this.

Though perhaps we can solve this by adding a `ShowFmt` to the tests? We could easily make the `Buildable` instances just display the first `n` bytes of each hash (or perhaps the prefix and suffix: `aaaa...aaaa`).

rvl commented on lib/core/src/Cardano/Wallet/Primitive/Types/TokenQuantity.hs:
```diff
@@ -0,0 +1,121 @@
+add :: TokenQuantity -> TokenQuantity -> TokenQuantity
+add (TokenQuantity x) (TokenQuantity y) = TokenQuantity $ x + y
+
+subtract :: TokenQuantity -> TokenQuantity -> TokenQuantity
```
Is subtract used anywhere?
And what happens if its result is negative?
I think this function should either be failable (i.e. Maybe TokenQuantity) or removed.

jonathanknowles replied:
@rvl I've now removed this function.

I think you're right that if we need to compare or find the difference between two quantities, there are other, safer methods that we can use. 👍🏻 

Fixed in 9a5bc7ea71dd4e7ab46b4e02f6b04d12badd0686.

rvl replied:
Thanks.

rvl reviewed (commented):
It looks nice, but there's a lot of fluff.
I am only half-way through this diff, and am getting the feeling there are many function defined which aren't really needed.

jonathanknowles committed:
Remove `subtract` operation from `TokenQuantity`.

Fixed in reponse to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2384#discussion_r540838958

jonathanknowles committed:
Use the `W` prefix within `Wallet.DB.Sqlite.TH` for wallet types.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2384#discussion_r540767462

jonathanknowles committed:
Clarify comment regarding SQLite and big integer support.

The original comment was correct, but the underlying reason is that SQLite
currently does not have a big integer type.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2384#discussion_r540768231

jonathanknowles committed:
Use function `readTxOutTokens` to simplify `selectTxs`.

The `readTxOutTokens` function fetches the complete set of tokens
associated with a `TxOut`.

We use this function to reduce duplication within `selectTxs`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2384#discussion_r540787527

jonathanknowles committed:
Use `map` instead of `<$>` in `txHistoryFromEntity`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2384#discussion_r540766261

jonathanknowles committed:
Make property tests run in parallel for token bundle type.

jonathanknowles committed:
Implement suggestions from `hlint`.

jonathanknowles committed:
Add strict `NonEmptyMap` type and tests.

This change introduces a `NonEmptyMap` type with strict semantics.

The implementation is based on the implementation of 'Data.Map.Strict'
provided by the 'containers' package, but provides an extra guarantee
that the map contains at least one entry at all times.

At the time of writing, there are other similar types available on GitHub, but
these implementations all extend the lazy `Data.Map` type, and thus provide
lazy semantics.

However, in the case of long-lived data structures, a strict type is more
useful, as it allows us to avoid building up large numbers of unevaluated
thunks.

rvl commented on lib/strict-non-empty-containers/src/Data/Map/NonEmpty/Strict/Internal.hs:
```diff
@@ -0,0 +1,190 @@
+-- Since this function may reduce the number of elements to zero, it returns
+-- an ordinary map.
+--
+delete :: Ord k => k -> NonEmptyMap k a -> Map k a
```
This seemed to me like it would be difficult to use.
I thought that `Maybe (NonEmptyMap)` would be more usable in general.
Checking how it is actually used (one instance) - the result is immediately changed to Maybe.
So I suggest changing the type of this function to match.

jonathanknowles replied:
I did initially consider the type you suggested, but then later on changed it to match the existing (lazy) non-empty map implementations out there.

But I agree that your type makes more sense in the context that we use it, so I'm happy to change it back.

Fixed in: d5a54cd64c53f2f09c8c4703bbbab08a68d037ec.

rvl commented on lib/strict-non-empty-containers/src/Data/Map/NonEmpty/Strict/Internal.hs:
```diff
@@ -0,0 +1,190 @@
+    , rest
+        :: !(Map.Map k v)
+    }
+    deriving (Eq, Foldable, Functor, Generic, Show, Traversable)
```
This representation seems to make everything harder.
What about wrapping a Map in newtype, and preserve your invariants in the functions which operate on the NonEmptyMap.

jonathanknowles replied:
> What about wrapping a Map in newtype, and preserve your invariants in the functions which operate on the NonEmptyMap.

That's an interesting idea. Would you mind if I experiment with that in a future PR?

The only (minor) downside I can think of is that we might have some "impossible" cases within pattern matches. But it might be simpler overall.

The published non-empty (lazy) map types all use the "least" element technique, which I borrowed here. (For example: [nonempty-containers](https://hackage.haskell.org/package/nonempty-containers-0.3.4.1/docs/src/Data.Map.NonEmpty.Internal.html#NEMap).)

rvl reviewed (approved):
Too fluffy for my liking, but otherwise good.
Just remember to address @KtorZ comments before merging.

jonathanknowles committed:
Remove unsafe constructors for `TokenName` and `TokenPolicyId`.

This change removes the smart constructors for the `TokenName` and
`TokenPolicyId` types.

In the case of `TokenPolicyId`, this constructor was actually unsafe.

The only place these functions were actually used was in the test code,
which has now been updated to use `fromText`.

In future, we can provide a function that produces a `TokenPolicyId`
value by hashing the content of a token policy.

Fixed in response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2384#discussion_r540810284

jonathanknowles committed:
Change the type of `NonEmptyMap.delete` to return `Maybe`.

Change the type of `NonEmptyMap.delete` from:

```
delete :: Ord k => k -> NonEmptyMap k a -> Map k a
```

to:

```
delete :: Ord k => k -> NonEmptyMap k a -> Maybe (NonEmptyMap k a)
```

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2384#discussion_r544840301

jonathanknowles committed:
Change the inner type of `TokenQuantity` to `Natural`.

At present, the only use case for these quantities is to be included in
token bundles held within transaction outputs, and these must never be
negative.

When we build support for minting and burning of tokens, we may wish to
parameterize this type and allow it to be instantiated with 'Integer'.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2384#discussion_r544243932

jonathanknowles committed:
Use longer qualifiers for imports of token types.

This change replaces imports of this style:

```
import qualified Primitive.Types.TokenBundle   as TB
import qualified Primitive.Types.TokenMap      as TM
import qualified Primitive.Types.TokenQuantity as TQ
```

With:

```
import qualified Primitive.Types.TokenBundle   as TokenBundle
import qualified Primitive.Types.TokenMap      as TokenMap
import qualified Primitive.Types.TokenQuantity as TokenQuantity
```

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2384#discussion_r544948654
